### PR TITLE
Fix out of bounds buffer access in backtrace

### PR DESCRIPTION
**Pull request by marcioapm**

### CODE REVIEW

**Comment by jacob-carlborg** [comment]
Is this possible to test?

**Comment by marcioapm** [comment]
> Is this possible to test?

yes, try this:

File: src/rt/backtrace/dwarf.d [hunk 1]
```diff
@@ -82,24 +82,30 @@ int traceHandlerOpApplyImpl(const void*[] callstack, scope int delegate(ref size
         char[1536] buffer = void; buffer[0] = 0;
         char[256] addressBuffer = void; addressBuffer[0] = 0;
 
-        if (locations.length > 0 && locations[i].line != -1)
+        if (locations.length > 0 && locations[i].line != -1) {
             snprintf(addressBuffer.ptr, addressBuffer.length, "%.*s:%d ", cast(int) locations[i].file.length, locations[i].file.ptr, locations[i].line);
-        else
+            addressBuffer[$-1] = '\0';
```
Comment (marcioapm) [inline src/rt/backtrace/dwarf.d]: Oh, what is this? Just noticed this is not needed, we are in Dland after all.

Comment (nemanja-boric-sociomantic) [reply src/rt/backtrace/dwarf.d]: This looks like it's needed, since you are passing the `addressBuffer.ptr` to sprintf?

**Comment by marcioapm** [comment]
I didn't think you meant automated test :) I suppose an automated test could be written for it, but I wouldn't know how to go about it. Could someone chip in? I would like to be done with this as soon as possible.

**Comment by jacob-carlborg** [comment]
There’s a `test` directory. Have a look there. I’m not that familiar with testing in druntime either.

**Comment by marcioapm** [comment]
There, I hope it works on all platforms.

File: src/rt/backtrace/dwarf.d [hunk 2]
```diff
@@ -90,12 +90,20 @@ int traceHandlerOpApplyImpl(const void*[] callstack, scope int delegate(ref size
         char[1024] symbolBuffer = void;
         int bufferLength;
         auto symbol = getDemangledSymbol(frameList[i][0 .. strlen(frameList[i])], symbolBuffer);
-        if (symbol.length > 0)
-            bufferLength = snprintf(buffer.ptr, buffer.length, "%s%.*s [0x%x]", addressBuffer.ptr, cast(int) symbol.length, symbol.ptr, callstack[i]);
-        else
-            bufferLength = snprintf(buffer.ptr, buffer.length, "%s[0x%x]", addressBuffer.ptr, callstack[i]);
+        if (symbol.length > 0) {
```
Comment (jacob-carlborg) [inline src/rt/backtrace/dwarf.d]: Curly braces should be on their own lines.

**Comment by marcioapm** [comment]
There.

**Comment by jacob-carlborg** [comment]
Should have a changelog entry (look in the changelog directory) or refer to an issue at issues.dlang.org. 

Please also squash the commits.

**Comment by jacob-carlborg** [comment]
> I have opened a new issue so it can be referenced here. Shall I also change the status to fixed?

No, that will be handled automatically. Technically it’s not fixed yet since it’s not merged into master 😉.

**Pull request merged**